### PR TITLE
`Programming exercises`: Prevent online code editor from overwriting other files during build/refresh

### PR DESCRIPTION
**Pull request by HawKhiem**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **LocalVC** and **Jenkins**.

### Motivation and Context
This PR fixes a critical race condition in the Monaco Editor that causes file content corruption during rapid file switching scenarios. The issue was reported in issue https://github.com/ls1intum/Artemis/issues/11353 and occurs when:

1. A user edits a file (e.g., BubbleSort.java) and submits changes
2. While the build is running, they switch to another file (e.g., MergeSort.java) and make changes
3. Switch back to the first file and wait till the build is finished
4. Switch to the other file containing unsubmitted changes (MergeSort) then switch back to the first file (BubbleSort)
5. Due to delayed text change emissions, the changes from the first file get applied to the currently selected file, causing content corruption

The root cause was that delayed text emissions were being applied to the currently selected file instead of the file they originated from, leading to data loss and file corruption.

### Short notice
I was able to reproduce the bug more consistently than described in the original issue. Here’s what I found:
1. The problem almost always occurs when the files are not syntax-highlighted.
2. It is unclear whether the bug also occurs if only one of the files is not syntax-highlighted.
3. The position of the files in the file tree does not matter; I have reproduced the bug even with files in the middle of the tree.
4. The file type also does not matter.
5. The bug occurs when a user edits one file while it is not syntax-highlighted, submits the changes, switches to another file that is also not syntax-highlighted, edits it, and then switches back to the first file.
6. When are files not syntax-highlighted? (No guarantee—based on personal observation):
- After submitting changes and reloading while the build is still running — the files are not syntax-highlighted for a short time.
- When the build has finished, reloading usually restores syntax highlighting.
- Files are also briefly not syntax-highlighted when first opening them in the editor.

#### Core Changes:
1. **New Event Signature**: Modified the `textChanged` event from emitting a simple `string` to emitting `{ text: string; fileName: string }` to include explicit file context
2. **File Path Resolution**: Added [extractFilePathFromModel()] in monaco-editor.component.ts to properly extract full file paths from Monaco Editor model URIs
3. **Contex-based File Handling**: Updated [onFileTextChanged()] in code-editor-monaco.component.ts to apply changes to the correct file based on the fileName property, not the currently selected file
#### Component Updates:
- **Monaco Editor Component**: Modified to emit file context with all text changes
- **Programming Component**: Modified to handle new event signature and apply changes to correct files
- **Custom Build Plan Components**: Updated both Aeolus and regular build plan components
- **Markdown Editor Component**: Adapted to extract text from new event object format

### Steps for Testing
Prerequisites:
- 1 Instructor
- 2 Students  
- 1 Programming Exercise with multiple files (e.g., BubbleSort.java, MergeSort.java)

#### Race Condition Testing:
1. Log in as a student and navigate to the programming exercise
2. Open the code editor and make changes to one file (e.g., BubbleSort.java)
3. Click Submit and immediately switch to (e.g., MergeSort.java) while build is running
4. Make changes to MergeSort.java and switch back to BubbleSort.java
5. After build completes, verify that:
   - BubbleSort.java contains only its original changes
   - MergeSort.java contains only its changes
   - No file content corruption occurs

#### Rapid File Switching Testing:
1. Open multiple files in the editor
2. Rapidly switch between files while typing in each
3. Verify that all changes are preserved and applied to correct files
4. Submit changes and verify persistence through the workflow

#### Edge Case Testing:
1. Test during build states when syntax highlighting may be disabled
2. Test file switching during editor resets
3. Verify changes persist through submission workflows
4. Test with files that have similar names or content

#### Exam Mode Testing
Prerequisites:
- 1 Instructor
- 2 Students  
- 1 Programming Exercise with multiple files (e.g., BubbleSort.java, MergeSort.java)

Testing steps are identical

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

#### Performance Review

- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- New Features
  - Editors (code & markdown) now include file name with text changes, enabling precise per-file updates and background edits.

- Bug Fixes
  - Fixed cases where edits were wrongly applied to the selected file instead of the intended file.

- Refactor
  - Standardized change events across editors to emit a consistent, file-aware payload for improved integration and reliability.

### CODE REVIEW

**Review by wasnertobias — APPROVED** [review]

File: src/main/webapp/app/shared/monaco-editor/monaco-editor.component.ts [hunk 1]
```diff
@@ -177,22 +177,46 @@ export class MonacoEditorComponent implements OnInit, OnDestroy {
...
+        this.textChangedEmitTimeouts.set(modelKey, timeoutId);
+    }
+
+    private extractFilePathFromModel(model: monaco.editor.ITextModel | null): string {
+        const path = model?.uri?.path ?? '';
+        if (!path) {
+            return '';
+        }
+        // Path format: /model/<editorId>/<full/file/path>
+        const parts = path.split('/').filter(Boolean);
```
Comment (tobias-lippert) [inline src/main/webapp/app/shared/monaco-editor/monaco-editor.component.ts:214]: Maybe add a comment that the boolean filter filters empty strings
